### ISSUE TITLE
itBit fetchBalance Error with signature

### ISSUE THREAD

**Issue by trentduncan** [ccxt/ccxt#3594]
- OS: High Sierra v10.13.6
- Node.js v8.9.4
- CCXT v1.17.88
- Exchange: itBit
- Method: fetchBalance

The Issue:

When I try calling the fetchBalance Method I get back a 401 unauthorized from itBit for having a bad signature. I know my credentials are valid because when I use a different library ([npm itbit](https://www.npmjs.com/package/itbit)) the response is fine. 

Code:
```
const ccxt = require("ccxt");

const getBalances = async () => {
  try {
    const client = new ccxt.itbit({
      apiKey: 'API_KEY',
      secret: 'API_SECRET',
      userId: 'USER_ID',
      verbose: true
    });

    const accounts = await client.fetchBalance();
    console.log(accounts);
    return;
  } catch (error) {
    console.log(error);
    return error;
  }
};

getBalances();
```
 
```
fetch:
 itbit GET https://api.itbit.com/v1/wallets?userId='USER_ID' 
Request:
 { Authorization: 'API_KEY:signature',
  'Content-Type': 'application/json',
  'X-Auth-Timestamp': '1533766553321',
  'X-Auth-Nonce': '1533766553321' }
Response: 
{ Error: itbit GET https://api.itbit.com/v1/wallets?userId='USER_ID' 401 Unauthorized {"code":10002,"description":"Your request lacked a valid signature, and could not be authenticated.","requestId":"REQUEST_ID"}
```
I was curious if anybody else has ran into this same issue. I'm assuming it is because the signature is being generated incorrectly in some portion of the sign method in itbit.js. I am currently looking into it, but if somebody could help me out that would be much appreciated. 

Thanks,
Trent

**Comment by kroitor** [comment]
That can be a nonce issue, have you tried with a new key?

**Comment by trentduncan** [comment]
Yes. Also when I go back and forth between the other library and ccxt the other library works consistently.

**Comment by kroitor** [comment]
@trentduncan ok, I'll look into it and will get back to you with a fix. One question, though: can you provide a testing key (zero balances, we don't need your money, just need to verify it's working properly)? Would be great if you could send it to igor@ccxt.trade. No worries, if not.

**Comment by kroitor** [comment]
@trentduncan one thing to clarify, have you tried with a new key in CCXT first?

**Comment by trentduncan** [comment]
Unfortunately, I don't have access to a testing key at this time. Thank you for your quick reply I appreciate it.

And yes I was working with CCXT before I even attempted the new library.

**Comment by kroitor** [comment]
> Unfortunately, I don't have access to a testing key at this time. Thank you for your quick reply I appreciate it.

No worries, I'll compare against the other library above. But let's clarify the nonce questions first, you've tried with a new API key with CCXT first, right?

**Comment by trentduncan** [comment]
@kroitor Yes I can confirm I tried ccxt with a fresh key before the other library.

**Comment by kroitor** [comment]
@frankhucek this is a different issue, you're not using proper rate-limiting, as documented in the Manual: 

- https://github.com/ccxt/ccxt/wiki/Manual#rate-limit
- https://github.com/ccxt/ccxt/wiki/Manual#authentication

Add `enableRateLimit: true` to your constructor params, that should solve it. Also make sure your nonce on the apiKey is not screwed by any other software.

**Comment by trentduncan** [comment]
@kroitor I have confirmed it is an issue in the sign method in itbit.js.

When I changed the signature generation in CCXT It is working.

Here is the code for reference:

ccxt/js/itbit.js

```
// this import is also required
const crypto = require('crypto');

sign( path, api = "public", method = "GET", params = {}, headers = undefined, body = undefined) {
    let url = this.urls["api"] + "/" + this.version + "/" + this.implodeParams(path, params);
    let query = this.omit(params, this.extractParams(path));
    if (method === "GET" && Object.keys(query).length)
      url += "?" + this.urlencode(query);
    if (method === "POST" && Object.keys(query).length) body = this.json(query);
    else body = "";
    if (api === "private") {
      this.checkRequiredCredentials();
      let nonce = this.nonce().toString();
      let timestamp = nonce;

      // start of new code

      const message =
        nonce +
        JSON.stringify([
          method,
          url,
          "",
          nonce.toString(),
          timestamp.toString()
        ]);

      const hashBuffer = crypto
        .createHash("sha256")
        .update(message)
        .digest();

      const bufferToHash = Buffer.concat([Buffer.from(url), hashBuffer]);

      const signature = crypto
        .createHmac("sha512", this.secret)
        .update(bufferToHash)
        .digest("base64");

      // end of new code

      headers = {
        Authorization: this.apiKey + ":" + signature,
        "Content-Type": "application/json",
        "X-Auth-Timestamp": timestamp,
        "X-Auth-Nonce": nonce
      };
    }
    return { url: url, method: method, body: body, headers: headers };
  }
```

I don't know if this change will work for all of the methods you support for itbit because the empty string is hardcoded since it is a GET request. But it fixes the fetchBalance issue.

**Comment by kroitor** [comment]
> Is there a plan to integrate this change into the library?

Yes, however, we will be happy to see contributions from you )

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-contribute-code
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#multilanguage-support
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes

**Comment by kroitor** [comment]
@gRittyKitty i've uploaded the fix provided by @trentduncan, it will be available in 5-10 minutes in version 1.18.113. It would be great if you could help us test it and report if everything works on your side. Thx!

**Comment by gRittyKitty** [comment]
Thx @kroitor 
So far so good. I updated and it is working on both Mac and ubuntu (not that I'd expect that to be a problem, but still). I plan to be using this for sending/cancelling orders, getting open orders, getting trades, balances. So seems like it ought to be a reasonable test. 
I'll plan on reporting issues. If you want positive (as in, yes its working), lmk where to post -- I did not see that sort of thing on the contributing pages.
